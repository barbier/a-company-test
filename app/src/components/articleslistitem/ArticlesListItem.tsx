import { useContext, useState } from 'react'
import {
  ListItem,
  ListItemText,
  ListItemTitle,
  ListItemButton,
} from '../../styles/styles'
import { IArticle } from '../../interfaces/Article'

const { ArticleContext } = require('../../contexts/context')

const ArticlesListItem: React.FC<{article: IArticle}> = ({article}) => {

  const [read, setRead] = useState(false)
  // eslint-disable-next-line
  const [selectedArticle, setSelectedArticle] = useContext(ArticleContext)

  const handleReadMoreClick = () => {
    setRead(true)
    setSelectedArticle(article)
  }
  
  return (
    <ListItem className={read ? 'read' : ''}>
      <ListItemTitle>{article.title}</ListItemTitle>
      <ListItemText>{article.description}</ListItemText>
      <ListItemButton onClick={() => handleReadMoreClick()}>
        Read more
      </ListItemButton>
    </ListItem>
  )
}

export default ArticlesListItem