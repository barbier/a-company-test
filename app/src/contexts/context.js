import { createContext } from 'react'

export const ArticleContext = createContext()
