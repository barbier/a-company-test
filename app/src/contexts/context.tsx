import React, { useState, createContext, useContext } from "react";
import { IArticle } from "../interfaces/Article";
import { IContext } from "../interfaces/ContextProvider";

const useArticle = (): any => {
  const [selectedArticle, setSelectedArticle] = useState({});

  return {
    selectedArticle,
    chooseArticle: (article: IArticle) => setSelectedArticle(article),
  };
};

const ArticleContext = createContext({} as any);

export const ArticleContextProvider: React.FC<{ children: any }> = ({
  children,
}: IContext) => (
  <ArticleContext.Provider value={useArticle()}>
    {children}
  </ArticleContext.Provider>
);

export const useSelectedArticle = (): any =>
  useContext(ArticleContext).selectedArticle;
export const useSetSelectedArticle = (article: IArticle): any =>
  useContext(ArticleContext).chooseArticle(article);
