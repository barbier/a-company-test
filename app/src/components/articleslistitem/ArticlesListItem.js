import { useContext, useState } from 'react'
import { ArticleContext } from '../../contexts/context'
import {
  ListItem,
  ListItemText,
  ListItemTitle,
  ListItemButton,
} from '../../styles/styles'

function ArticlesListItem(props) {
  const { article } = props
  const [read, setRead] = useState(false)
  // eslint-disable-next-line
  const [selectedArticle, setSelectedArticle] = useContext(ArticleContext)

  const handleReadMoreClick = () => {
    setRead(true)
    setSelectedArticle(article)
  }
  
  return (
    <ListItem className={read ? 'read' : ''}>
      <ListItemTitle>{article.title}</ListItemTitle>
      <ListItemText>{article.description}</ListItemText>
      <ListItemButton onClick={() => handleReadMoreClick()}>
        Read more
      </ListItemButton>
    </ListItem>
  )
}

export default ArticlesListItem