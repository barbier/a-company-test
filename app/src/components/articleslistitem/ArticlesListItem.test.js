import { render, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { ArticleContext } from '../../contexts/context'
import ArticlesListItem from './ArticlesListItem'

describe("Test Articles List rendering options", () => {
  let selectedArticle
  let setSelectedArticle
  let article
  beforeEach(() => {
    selectedArticle = {}
    setSelectedArticle = () => {}
    article = {
      "source": {
        "id": "cnbc",
        "name": "CNBC"
      },
      "author": "Amelia Lucas",
      "title": "Coca-Cola stock jumps as strong sales of Coke Zero Sugar continue to drive revenue growth - CNBC",
      "description": "Coca-Cola on Friday reported quarterly earnings that met analysts' expectations.",
      "url": "https://www.cnbc.com/2019/10/18/coca-cola-ko-earnings-q3-2019.html",
      "urlToImage": "https://image.cnbcfm.com/api/v1/image/105375236-1533317591071rtx6drjw.jpg?v=1571332846",
      "publishedAt": "2019-10-18T10:42:46Z",
      "content": "Coca-Cola on Friday reported quarterly revenue that topped analysts' expectations as more customers are drawn in by healthier options, like Zero Sugar soda and smaller size cans. \r\nShares of the company jumped 1.6% in premarket trading.\r\n\"Our performance giveâ€¦ [+1887 chars]"
    }
  })
  
  test("It should match snapshot", () => {
    
    const component = renderer.create(
      <ArticleContext.Provider value={[selectedArticle, setSelectedArticle]}>
        <ArticlesListItem article={article} />
      </ArticleContext.Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("It shouldn't have any class names on render", () => {
    const {container} = render(
      <ArticleContext.Provider value={[selectedArticle, setSelectedArticle]}>
        <ArticlesListItem article={article} />
      </ArticleContext.Provider>
    )
    const listItem = container.querySelector(".read")
    
    expect(listItem).toBeFalsy()
  })

  test('It should have any class names "read" after clicking on button', () => {
    const {container} = render(
      <ArticleContext.Provider value={[selectedArticle, setSelectedArticle]}>
        <ArticlesListItem article={article} />
      </ArticleContext.Provider>
    )
    const button = container.querySelector('button')

    fireEvent.click(button)

    const listItem = container.querySelector(".read")
    
    expect(listItem).toBeTruthy()
  })
})