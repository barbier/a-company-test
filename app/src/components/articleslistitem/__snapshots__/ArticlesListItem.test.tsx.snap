// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Articles List rendering options It should match snapshot 1`] = `
<li
  className=" css-17vaqjz"
>
  <h2
    className="css-ybopvn"
  >
    Coca-Cola stock jumps as strong sales of Coke Zero Sugar continue to drive revenue growth - CNBC
  </h2>
  <p
    className="css-jd954d"
  >
    Coca-Cola on Friday reported quarterly earnings that met analysts' expectations.
  </p>
  <button
    className="css-oxpfok"
    onClick={[Function]}
  >
    Read more
  </button>
</li>
`;
